compiler/rustc_metadata/src/locator.rs

```diff
@@ -706,54 +709,57 @@ impl<'a> CrateLocator<'a> {
+            // e.g. symbolic links. If we canonicalize too early, we resolve
+            // the symlink, the file type is lost and we might treat rlibs and
+            // rmetas as dylibs.
+            let Some(file) = loc_orig.file_name().and_then(|s| s.to_str()) else {
+                return Err(CrateError::ExternLocationNotFile(self.crate_name, loc_orig.clone()));
             };
-
-            if file.starts_with("lib") && (file.ends_with(".rlib") || file.ends_with(".rmeta"))
-                || file.starts_with(self.target.dll_prefix.as_ref())
-                    && file.ends_with(self.target.dll_suffix.as_ref())
-            {
-                // Make sure there's at most one rlib and at most one dylib.
-                // Note to take care and match against the non-canonicalized name:
-                // some systems save build artifacts into content-addressed stores
-                // that do not preserve extensions, and then link to them using
-                // e.g. symbolic links. If we canonicalize too early, we resolve
-                // the symlink, the file type is lost and we might treat rlibs and
-                // rmetas as dylibs.
-                let loc_canon = loc.canonicalized().clone();
-                let loc = loc.original();
-                if loc.file_name().unwrap().to_str().unwrap().ends_with(".rlib") {
-                    rlibs.insert(loc_canon, PathKind::ExternFlag);
-                } else if loc.file_name().unwrap().to_str().unwrap().ends_with(".rmeta") {
-                    rmetas.insert(loc_canon, PathKind::ExternFlag);
-                } else {
-                    dylibs.insert(loc_canon, PathKind::ExternFlag);
+            // FnMut cannot return reference to captured value, so references
+            // must be taken outside the closure.
+            let rlibs = &mut rlibs;
+            let rmetas = &mut rmetas;
+            let dylibs = &mut dylibs;
+            match (|| {
+                if file.starts_with("lib") {
+                    if file.ends_with(".rlib") {
+                        return Some(rlibs);
+                    }
+                    if file.ends_with(".rmeta") {
+                        return Some(rmetas);
+                    }
+                }
+                let dll_prefix = self.target.dll_prefix.as_ref();
+                let dll_suffix = self.target.dll_suffix.as_ref();
+                if file.starts_with(dll_prefix) && file.ends_with(dll_suffix) {
+                    return Some(dylibs);
+                }
+                None
+            })() {
+                Some(dst) => {
+                    dst.insert(loc_canon.clone(), PathKind::ExternFlag);
+                }
+                None => {
+                    self.crate_rejections
+                        .via_filename
+                        .push(CrateMismatch { path: loc_orig.clone(), got: String::new() });
                 }
```

Comment:
I am personally not a fan of matching on a closure, maybe something like this?
```rust
            'accept_crate: {
                if file.starts_with("lib") {
                    if file.ends_with(".rlib") {
                        rlibs.insert(loc_canon.clone(), PathKind::ExternFlag);
                        break 'accept_crate;
                    }
                    if file.ends_with(".rmeta") {
                        rmetas.insert(loc_canon.clone(), PathKind::ExternFlag);
                        break 'accept_crate;
                    }
                }
                let dll_prefix = self.target.dll_prefix.as_ref();
                let dll_suffix = self.target.dll_suffix.as_ref();
                if file.starts_with(dll_prefix) && file.ends_with(dll_suffix) {
                     dylibs.insert(loc_canon.clone(), PathKind::ExternFlag);
                     break 'accept_crate;
                }
                
                self.crate_rejections
                        .via_filename
                        .push(CrateMismatch { path: loc_orig.clone(), got: String::new() });
           };
```

Reply:
That's fragile because removing any `break` would produce incorrect logic. I've assigned the result to a local variable, do you like that better?